Pausing before turning sysproxy off

oxtoacart opened the pull request:
For getlantern/lantern-internal#1121

For the sysproxy `off` command, these changes allow the parent process to execute sysproxy-cmd immediately and then release it to actually turn the system proxy off by sending a character on stdin or just closing stdin. On Windows, the `off` command will also complete in the event of a system shutdown.

joesis commented on common.h:
```diff
@@ -1,5 +1,6 @@
 #include <errno.h>
 #include <stdbool.h>
+#include <stdio.h>
```
Is it necessary?

joesis commented:
Maybe document the new kinda-unexpected behaviour in Makefile? One would be suprised that `sysproxy-cmd off ...` doesn't return immediately, and `^C` behaves differently from `^D` on non-windows platforms.

joesis commented:
Or alternatively, add an option `-cleanup` to `sysproxy-cmd on host port` to wait and clean up? Let me have a try.

oxtoacart commented on main.c:
```diff
@@ -1,14 +1,36 @@
+void turnOffProxyOnSignal(int signal)
+{
+  toggleProxy(false, proxyHost, proxyPort);
+  exit(0);
```
@joesis I added this to make sure that the process exits.

oxtoacart commented on main.c:
```diff
@@ -1,14 +1,36 @@
   exit(INVALID_FORMAT);
 }
 
+void turnOffProxyOnSignal(int signal)
```
I renamed this to be more descriptive.

joesis commented:
Thanks @oxtoacart! Merging.

oxtoacart commented:
@joesis Unfortunately, this doesn't work on Windows anymore after the refactoring. I confirmed that the version from before the refactoring still works. I'm investigating now.

oxtoacart commented:
To be specific, the behavior on stopping Lantern cleanly works, but the system shutdown handling isn't working.

oxtoacart commented:
Interestingly, the program itself seems to work correctly, but when I run it from within Lantern it does not ...

oxtoacart commented:
Hmm, it works now. I think I must have made a mistake with the build or something.

joesis commented:
Maybe my method fails sometimes? I was trying to test flashlight but my Internet connection is terrible at this time of the day to even run `glide`.

joesis commented:
Hmm I can reproduce it too. After manually changed flashlight to not clear proxy setting at start up, the proxy is still set after system reboot. Running more tests...

joesis commented:
Ok it's caused by `Unable to finish C:\Users\IEUser\AppData\Roaming\byteexec\sysproxy-cmd.exe: exit status -1073740771`

joesis commented:
`-1073740771` equals to 0xC000041D(STATUS_FATAL_USER_CALLBACK_EXCEPTION)

joesis commented:
Above error is from Windows 10. The very same flashlight binary works on XP, and on Win7 32bit, the exit status becomes 1073807364(0x40010004). It's definitely something with the code I added.

oxtoacart commented:
Ithink I fixed it, will check in soon.

On Mon, Aug 28, 2017 at 9:35 AM joesis <notifications@github.com> wrote:

> Above error is from Windows 10. The very same flashlight binary works on
> XP, and on Win7 32bit, the exit status becomes 1073807364(0x40010004). It's
> definitely something with the code I added.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/getlantern/sysproxy-cmd/pull/7#issuecomment-325371404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AExNzvdd2jwIniYZDlV62IGkGTsBZ6bbks5sctBJgaJpZM4PDlHT>
> .
>
-- 
Cheers,
Ox

PGP Key - https://pgp.mit.edu/pks/lookup?op=get&search=ox@getlantern.org

-----------------------------------------------------------------------------------------

"I love people who harness themselves, an ox to a heavy cart,
who pull like water buffalo, with massive patience,
who strain in the mud and the muck to move things forward,
who do what has to be done, again and again."

- Marge Piercy

joesis commented:
Weird, I still get exit status `0x40010004` on Win 7 32bit and a new exit status `0xC000026B`(STATUS_DLL_INIT_FAILED_LOGOFF) on Win10 64bit when unsetting system proxy. `sysproxy-cmd` has correct shasum in both cases.

oxtoacart commented:
@joesis I get that same error, but the proxy unsets successfully. Do you see it unsetting successfully?

oxtoacart commented:
I should mention that I tested this by temporarily disabling the unset on startup functionality. Even with a production build, you can test by immediately browsing to whatismyipaddress.com after logging in or opening chrome://net-internals#proxy, as there is a noticeably delay before the unset on startup functionality runs.

joesis commented:
Yes @oxtoacart I changed the unset on startup branch to exit immediately. On another Win10 VM, I got `0x40010004` but the proxy was unset successfully. The Win7 and Win10 box I mentioned above didn't unset proxy during OS reboot.

oxtoacart commented:
@joesis Weird. Do you see the same issue with the beta build from here?

https://raw.githubusercontent.com/getlantern/lantern-binaries/master/lantern-installer-preview.exe

oxtoacart commented:
I had only been testing with logoff, so I tested a restart just to be safe. It still worked.

I'm on Windows 10 Pro Insider Preview Build `16251.rs3_release.170721-2122`

joesis commented:
@oxtoacart With the preview build, the proxy can be cleared in both VMs, but I'm not quite sure if it's reset on shutdown or startup. Anyway, things work better than before with the sysproxy changes, if not perfect.

The `0x40010004` status is expected actually, as `sysproxy-cmd.exe` can receive the shutdown message earlier and quit before `flashlight`.

I'll check more on this as I can reproduce.

oxtoacart commented:
Thanks @joesis! Something I do to test with the preview build is to enable Lantern on system startup. What I observe is that for about 1 minute or so after logging in, whatismyipaddress.com shows my local address. However, once the proxy kicks in, it shows my proxy's address. If unsetting didn't work on shutdown, rather than showing my local address it would show an error until Lantern is back online.
